pennylane/device_tracker.py

```diff
@@ -0,0 +1,171 @@
+
+    Note that with backpropagation, this functions should take ``qnode.device``
+    instead of the device used to create the QNode.
+
+    .. code-block:: python
+
+        dev = qml.device('default.qubit', wires=1)
+
+        @qml.qnode(dev, diff_method="parameter-shift")
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+    With the default version, total execution information is printed on
+    each device execution.  The printed data depends on the device and tracker version,
+    but for standard PennyLane devices, the object will track executions and shots.
+
+    >>> with qml.track(circuit.device) as tracker:
+    ...    qml.grad(circuit)(0.1, shots=10)
+    Total: executions = 1	shots = 10
+    Total: executions = 2	shots = 20
+    Total: executions = 3	shots = 30
+
+    In with the ``'timing'`` implementation, the instance also tracks the time
+    between entering the context and the completion of an execution.
+
+    >>> with qml.track(circuit.device, version='timing') as timing_tracker:
+    ...    circuit(0.1)
+    ...    circuit(0.2)
+    Total: executions = 1	time = 0.0011134147644042969
+    Total: executions = 2	time = 0.0027322769165039062
+
+    After completion, one can also access the recorded information:
+
+    >>> timing_tracker.totals
+    defaultdict(int, {'executions': 2, 'shots': 30, 'time': 0.00311279296875})
+    >>> timing_tracker.history
+    defaultdict(list,
+            {'executions': [1, 1],
+             'shots': [None, None],
+             'time': [0.0012764930725097656, 0.0018362998962402344]})
+
+    By specifying ``reset_on_enter=False``, you can reuse the same tracker accross
+    multiple runtime contexts.
+
+    >>> with qml.track(circuit.device, reset_on_enter=False) as tracker:
+    ...     circuit(0.1)
+    Total: executions = 1
+    >>> with tracker:
+    ...     circuit(0.2)
+    Total: executions = 2
+
+    """
+    if version == "timing":
+        return TimingTracker(dev, **kwargs)
+    elif version == "default":
+        return DefaultTracker(dev, **kwargs)
+    else:
+        raise qml.QuantumFunctionError(
+            f"version {version} supplied to track. " f"Current options are `timing` and `default`."
+        )
+
+
+class DefaultTracker:
+    """
+    Class docstring
+    """
+
+    def __init__(self, dev=None, reset_on_enter=True):
+        """
+        docstring
+        """
+        self.reset_on_enter = reset_on_enter
+
+        self.reset()
+        self.tracking = False
+
+        if dev is not None:
+            dev.tracker = self
+
+    def __enter__(self):
+        """
+        docstring for enter
+        """
+        if self.reset_on_enter:
+            self.reset()
+
+        self.tracking = True
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        """
+        docstring for exit
+        """
+        self.tracking = False
+
+    def update_and_record(self, **current):
```

Comment:
I would suggest including these two lines at relevant places rather than having a function for this. Strictly speaking, this function does two things (as per the name having the `and` word in there, and I'd think that having the two units of `update` and `record` be separately considered and tested should also help down the line.